### PR TITLE
collective rework and initial rxm integration

### PR DESCRIPTION
**Pull request by ooststep**
This PR slightly reworks prov/util's collective implementation and does an initial integration with rxm provider for join and barrier.

### CODE REVIEW

File: prov/util/src/util_coll.c [hunk 1]
```diff
@@ -856,28 +847,27 @@ int ofi_av_set(struct fid_av *av, struct fi_av_set_attr *attr,
 
 ssize_t ofi_ep_barrier(struct fid_ep *ep, fi_addr_t coll_addr, void *context)
 {
-	struct util_coll_mc *coll_mc = (struct util_coll_mc *) coll_addr;
+	struct util_coll_mc *coll_mc = (struct util_coll_mc *)coll_addr;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: Cast integers (fi_addr_t) to pointers using (uintptr_t)

File: prov/util/src/util_coll.c [hunk 2]
```diff
@@ -673,37 +674,29 @@ int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
...
 		util_coll_init_cid(util_coll_cid);
 		util_coll_cid_initialized = 1;
 	}
 
-	new_coll_mc->mc_fid.fid.fclass = FI_CLASS_MC;
-	new_coll_mc->mc_fid.fid.context = context;
-	new_coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
-	new_coll_mc->mc_fid.fi_addr = (uintptr_t) new_coll_mc;
-
-	new_coll_mc->ep = ep;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: The changes surrounding this area are unrelated to writing an entry to the CQ.  Changes fall into wrong patch?

File: include/ofi_coll.h [hunk 3]
```diff
@@ -111,16 +111,17 @@ struct util_coll_comp_item;
...
 };
 
 struct util_coll_comp_item {
 	struct util_coll_hdr	hdr;
 	enum util_coll_op_type	op_type;
-	void			*data;
-	util_coll_comp_t	comp_fn;
+	void					*op_context;
+	void					*data;
+	util_coll_comp_t		comp_fn;
```
Comment (shefty) [inline include/ofi_coll.h]: nit: spacing looks off

File: prov/util/src/util_coll.c [hunk 4]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
 {
+	struct util_coll_mc *coll_mc;
+	struct slist_entry *entry;
 	struct util_coll_hdr *hdr;
 	struct util_coll_xfer_item *xfer_item;
+	struct iovec iov;
+	struct fi_msg_tagged msg = { 0 };
+	int err;
+	coll_mc = ep->av->coll_mc;
+	if (coll_mc) {
```
Comment (shefty) [inline prov/util/src/util_coll.c]: reverse if check and return, so the entire function isn't indented an extra tab

File: prov/util/src/util_coll.c [hunk 5]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
+int ofi_coll_process_pending(struct util_ep *ep)
 {
+	struct util_coll_mc *coll_mc;
+	struct slist_entry *entry;
 	struct util_coll_hdr *hdr;
 	struct util_coll_xfer_item *xfer_item;
+	struct iovec iov;
+	struct fi_msg_tagged msg = { 0 };
+	int err;
+	coll_mc = ep->av->coll_mc;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: nit: please add blank line after variable declarations

File: prov/util/src/util_coll.c [hunk 6]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
+						  FI_COLLECTIVE);
+				if (err) {
+					return err;
+				}
+				break;
+			default:
+				break;
+			}
+			if (hdr->is_barrier &&
+			    !slist_empty(&coll_mc->barrier_list)) {
```
Comment (shefty) [inline prov/util/src/util_coll.c]: The slist_empty check isn't needed.  We always want to break on barrier (which is what the code will do, just with more checks).

File: prov/util/src/util_coll.c [hunk 7]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
+				do {
+					err = fi_tsendmsg(coll_mc->ep, &msg,
+							  FI_COLLECTIVE);
+					if (err && err != -EAGAIN) {
+						slist_insert_head(
+							&xfer_item->hdr.entry,
+							&coll_mc->deferred_list);
+						return err;
+					}
+				} while (err == -EAGAIN);
```
Comment (shefty) [inline prov/util/src/util_coll.c]: We can spin on EAGAIN without attempting to progress the EP, for example, by reading the CQ.  In this case, we should just keep the pending transfer pending.

File: prov/util/src/util_coll.c [hunk 8]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
+				iov.iov_base = xfer_item->buf;
+				iov.iov_len = (xfer_item->count *
+					       ofi_datatype_size(
+						       xfer_item->datatype));
+				msg.msg_iov = &iov;
+				msg.iov_count = 1;
+				msg.addr = coll_mc->av_set->fi_addr_array
+						   [xfer_item->dest_rank];
+				msg.tag = hdr->tag;
+				msg.context = (void *)coll_mc;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: nit: casts to void * aren't necessary

File: prov/util/src/util_coll.c [hunk 9]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
+			hdr = container_of(entry, struct util_coll_hdr, entry);
+			switch (hdr->type) {
+			case UTIL_COLL_SEND:
+				xfer_item = (struct util_coll_xfer_item *)hdr;
+				iov.iov_base = xfer_item->buf;
+				iov.iov_len = (xfer_item->count *
+					       ofi_datatype_size(
+						       xfer_item->datatype));
+				msg.msg_iov = &iov;
+				msg.iov_count = 1;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: The initialization of msg.msg_iov and msg.iov_count can be moved outside of all loops.

File: prov/util/src/util_coll.c [hunk 10]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
 
-static int util_coll_process_work_items(struct util_coll_mc *coll_mc)
+int ofi_coll_process_pending(struct util_ep *ep)
 {
+	struct util_coll_mc *coll_mc;
+	struct slist_entry *entry;
 	struct util_coll_hdr *hdr;
 	struct util_coll_xfer_item *xfer_item;
+	struct iovec iov;
+	struct fi_msg_tagged msg = { 0 };
```
Comment (shefty) [inline prov/util/src/util_coll.c]: please don't over-initialize data structures.  Most of these fields will be set further down.

File: prov/util/src/util_coll.c [hunk 11]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
+	int err;
+	coll_mc = ep->av->coll_mc;
+	if (coll_mc) {
+		while (!slist_empty(&coll_mc->pending_xfer_list)) {
+			entry = slist_remove_head(&coll_mc->pending_xfer_list);
+			hdr = container_of(entry, struct util_coll_hdr, entry);
+			switch (hdr->type) {
+			case UTIL_COLL_SEND:
+				xfer_item = (struct util_coll_xfer_item *)hdr;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: I see this in the original code, but we'll want to do something better here than cast two different structures to each other.

File: prov/util/src/util_coll.c [hunk 12]
```diff
@@ -557,77 +558,113 @@ static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
...
-			if (ret) {
-				return ret;
-			}
-
+			slist_insert_tail(entry, &coll_mc->pending_xfer_list);
 			slist_insert_tail(entry, &coll_mc->barrier_list);
 			break;
 		case UTIL_COLL_REDUCE:
-			reduce_item = (struct util_coll_reduce_item *) hdr;
+			reduce_item = (struct util_coll_reduce_item *)hdr;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: Don't hate spaces.  :)

File: prov/util/src/util_coll.c [hunk 13]
```diff
@@ -636,8 +673,7 @@ static int util_coll_process_work_items(struct util_coll_mc *coll_mc)
 			break;
 		}
 
-		if (hdr->is_barrier &&
-		    !slist_empty(&coll_mc->barrier_list)) {
+		if (hdr->is_barrier && !slist_empty(&coll_mc->barrier_list)) {
```
Comment (shefty) [inline prov/util/src/util_coll.c]: barrier check is sufficient

File: prov/rxm/src/rxm.h [hunk 14]
```diff
@@ -53,6 +53,7 @@
 #include <ofi_list.h>
 #include <ofi_proto.h>
 #include <ofi_iov.h>
+#include <ofi_coll.h>
```
Comment (shefty) [inline prov/rxm/src/rxm.h]: There aren't changes to this file that would require adding this include.

File: prov/rxm/src/rxm_cq.c [hunk 15]
```diff
@@ -597,13 +597,34 @@ ssize_t rxm_cq_handle_rndv(struct rxm_rx_buf *rx_buf)
...
+	return FI_SUCCESS;
+}
+
+static inline
+ssize_t rxm_cq_tx_handle_coll(struct rxm_tx_eager_buf *eager_buf)
+{
+	ofi_coll_handle_comp(eager_buf->pkt.hdr.tag,
+			      eager_buf->app_context);
+	return FI_SUCCESS;
+}
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: Both of the above functions can be void.  Neither return a useful value.

File: prov/rxm/src/rxm_cq.c [hunk 16]
```diff
@@ -597,13 +597,34 @@ ssize_t rxm_cq_handle_rndv(struct rxm_rx_buf *rx_buf)
...
+
 static inline
 ssize_t rxm_cq_handle_eager(struct rxm_rx_buf *rx_buf)
 {
 	uint64_t done_len = ofi_copy_to_iov(rx_buf->recv_entry->rxm_iov.iov,
 					    rx_buf->recv_entry->rxm_iov.count,
 					    0, rx_buf->pkt.data,
 					    rx_buf->pkt.hdr.size);
+	if(OFI_TAG_COLLECTIVE & rx_buf->pkt.hdr.tag) {
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: reverse logic

File: prov/rxm/src/rxm_cq.c [hunk 17]
```diff
@@ -1108,7 +1129,11 @@ static ssize_t rxm_cq_handle_comp(struct rxm_ep *rxm_ep,
 	case RXM_TX:
 		tx_eager_buf = comp->op_context;
 		assert(comp->flags & FI_SEND);
-		ret = rxm_finish_eager_send(rxm_ep, tx_eager_buf);
+		if (OFI_TAG_COLLECTIVE & tx_eager_buf->pkt.hdr.tag) {
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: reverse logic

File: prov/rxm/src/rxm_ep.c [hunk 18]
```diff
@@ -468,6 +470,19 @@ static int rxm_getname(fid_t fid, void *addr, size_t *addrlen)
...
+			struct fid_mc **mc, void *context)
+{
+	if (flags & FI_COLLECTIVE) {
+		struct fi_collective_addr *c_addr =
+			(struct fi_collective_addr *)addr;
+		return ofi_join_collective(ep, c_addr->coll_addr, c_addr->set,
+					   flags, mc, context);
+	} else {
+		return -FI_ENOSYS;
+	}
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c]: nit: move flag check to top of function and abort.  Outdent the rest of the code.

File: include/ofi_coll.h [hunk 19]
```diff
@@ -104,52 +108,68 @@ struct util_coll_reduce_item {
...
+struct util_coll_comp_data {
 	uint64_t		cid_buf[OFI_CONTEXT_ID_SIZE];
 	uint64_t		tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
 };
 
+struct util_coll_state {
+	struct dlist_entry	entry;
+	struct util_coll_mc	*coll_mc;
+	int			complete;
+};
```
Comment (shefty) [inline include/ofi_coll.h]: Does this need to be a separate structure, or can the fields just be added directly to util_coll_comp_item, or whatever structure is tracking the overall collective operation?

File: prov/rxm/src/rxm_cq.c [hunk 20]
```diff
@@ -1429,13 +1455,17 @@ void rxm_ep_do_progress(struct util_ep *util_ep)
...
+	ofi_coll_ep_progress(&util_ep->ep_fid);
 }
 
 void rxm_ep_progress(struct util_ep *util_ep)
 {
 	ofi_ep_lock_acquire(util_ep);
 	rxm_ep_do_progress(util_ep);
 	ofi_ep_lock_release(util_ep);
+
+	ofi_coll_process_pending(&util_ep->ep_fid);
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: We'll need to see how to handle progress, so that when collective support isn't requested (common case), we don't incur the overhead of checking empty lists and acquiring extra locks.

File: prov/rxm/src/rxm_cq.c [hunk 21]
```diff
@@ -1429,13 +1455,17 @@ void rxm_ep_do_progress(struct util_ep *util_ep)
 					     deferred_conn_entry, conn_entry_tmp)
 			rxm_ep_progress_deferred_queue(rxm_ep, rxm_conn);
 	}
+
+	ofi_coll_ep_progress(&util_ep->ep_fid);
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: See next comment below.

File: prov/util/src/util_coll.c [hunk 22]
```diff
@@ -249,7 +239,7 @@ static int util_coll_sched_recv(struct util_coll_mc *coll_mc, int src,
 	xfer_item->hdr.type = UTIL_COLL_RECV;
 	xfer_item->hdr.is_barrier = is_barrier;
 
-	xfer_item->hdr.tag = (recv_tag << 32) | (tag & 0xffffffff);
+	xfer_item->hdr.tag = (src_rank << 32) | (tag & (0xffffffff | OFI_TAG_COLL));
```
Comment (shefty) [inline prov/util/src/util_coll.c]: nit: I would restructure this to avoid making it look like OFI_TAG_COLL is associated with the lower 32-bits.  E.g.
```
OFI_TAG_COLL | (src_rank << 32) | (tag & 0xffffffff)
```
I would even look at removing the manipulation of the tag here completely, so that we just do
```
hdr.tag = tag  /* or htonl(tag) */
```
and fold in the src_rank value when we get the tag.

File: prov/util/src/util_coll.c [hunk 23]
```diff
@@ -330,7 +323,7 @@ static inline uint64_t util_coll_get_next_tag(struct util_coll_mc *coll_mc)
 
 	tag = tag << 16 | coll_mc->cid;
 	tag = tag << 16 | coll_mc->tag_seq++;
-	return tag;
+	return tag | OFI_TAG_COLL;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: I realize this is existing code, but it might probably be cleaner if collective tag (id?) were defined as a u32 and only used in that way.  Rename OFI_TAG_COLL to OFI_COLL_TAG_FLAG.

The use of tag to mean different things is getting confusing.  Particularly, when the entire tag is formed in separate locations.

File: prov/util/src/util_coll.c [hunk 24]
```diff
@@ -526,194 +519,129 @@ void util_coll_join_comp(struct util_coll_mc *coll_mc,
...
+	if (ofi_cq_write(ep->tx_cq, comp->op_context, FI_COLLECTIVE,
+			 sizeof(struct util_coll_comp_data), comp->data, 0,
+			 comp->hdr.tag)) {
 		FI_WARN(ep->domain->fabric->prov, FI_LOG_DOMAIN,
-			"barrier collective - eq write failed\n");
-*/
+			"barrier collective - cq write failed\n");
+	}
+	free(comp->data);
+	comp->state->complete = 1;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: The collective operation is done here.  Why isn't comp just freed?

File: prov/util/src/util_coll.c [hunk 25]
```diff
@@ -526,194 +519,129 @@ void util_coll_join_comp(struct util_coll_mc *coll_mc,
...
-			msg.iov_count = 1;
-			msg.addr = coll_mc->av_set->fi_addr_array[xfer_item->src_rank];
 			msg.tag = hdr->tag;
-			msg.context = (void *) coll_mc;
-
-			ret = fi_trecvmsg(coll_mc->ep, &msg, FI_COLLECTIVE);
-			if (ret) {
-				return ret;
+			msg.addr = coll_mc->av_set
+					   ->fi_addr_array[xfer_item->src_rank];
```
Comment (shefty) [inline prov/util/src/util_coll.c]: nit: it's okay to have lines go over 80 characters if it keeps the code more readable.

File: prov/util/src/util_coll.c [hunk 26]
```diff
@@ -526,194 +519,129 @@ void util_coll_join_comp(struct util_coll_mc *coll_mc,
...
-					slist_insert_head(&xfer_item->hdr.entry,
-							  &coll_mc->deferred_list);
-					return ret;
-				}
-			} while (ret == -EAGAIN);
-
-			slist_insert_tail(entry, &coll_mc->barrier_list);
+			msg.addr =
+				coll_mc->av_set
+					->fi_addr_array[xfer_item->dest_rank];
```
Comment (shefty) [inline prov/util/src/util_coll.c]: see below comment.

File: prov/util/src/util_coll.c [hunk 27]
```diff
@@ -760,14 +687,104 @@ static int util_coll_schedule(struct util_coll_mc *coll_mc)
 	return FI_SUCCESS;
 }
 
+int util_coll_state_match_complete(struct dlist_entry *entry, const void *arg)
```
Comment (shefty) [inline prov/util/src/util_coll.c]: static

File: prov/util/src/util_coll.c (hunk 22 above)
Comment (ooststep) [reply prov/util/src/util_coll.c]: the intent was to just mask the 31 bits that are the 'source' rank of the message rather than forcing the 'collective bit' to be set.  In that case, `OFI_TAG_COLL` is actually associated with the lower 32-bits in that together (bitwise OR) they form the rank mask `0x80000000FFFFFFFF`.  A bitwise AND with this mask would 'zero out' the rank part of the tag, but not modify the 'collective flag', cid, or seq parts of the tag.

Comment (shefty) [reply prov/util/src/util_coll.c]: I'm confused, unless 'tag' has the upper bit set, this clears the OFI_TAG_COLL flag.  If you just want to clear the source rank bits, I would do that more explicitly, then set them.  E.g.
```
tag = tag & ~OFI_TAG_SOURCE_BITS;
tag |= (src_rank << OFI_TAG_SOURCE_SHIFT)
```
which you could put into a static inline call if it's needed elsewhere.

**Comment by shefty** [comment]
Picked the first 2 commits (with minor updates) into PR #5321, so we can start integrating some of the changes.

File: prov/util/src/util_coll.c [hunk 28]
```diff
@@ -520,23 +521,17 @@ void util_coll_join_comp(struct util_coll_mc *coll_mc,
...
-			 &entry, (size_t) bytes, FI_COLLECTIVE) < 0)
+	if (ofi_cq_write(ep->tx_cq, comp->op_context, FI_COLLECTIVE,
+			 sizeof(struct util_coll_comp_data), comp->data, 0,
+			 comp->hdr.tag)) {
 		FI_WARN(ep->domain->fabric->prov, FI_LOG_DOMAIN,
-			"barrier collective - eq write failed\n");
-*/
+			"barrier collective - cq write failed\n");
+	}
+	free(comp->data);
```
Comment (shefty) [inline prov/util/src/util_coll.c]: There's no great way to comment on the commit messages, but the message is too terse.  As a nit, the preferred way of labeling commits would be:
```
provider/topic: Answer this commit will do what here
```
This is a good source describing commit message format:

https://chris.beams.io/posts/git-commit/

File: prov/util/src/util_coll.c [hunk 29]
```diff
@@ -474,13 +475,13 @@ void util_coll_join_comp(struct util_coll_mc *coll_mc,
...
 	struct fi_eq_err_entry entry;
-	struct util_coll_join_comp_data *comp_data;
+	struct util_coll_comp_data *comp_data;
 	struct util_ep *ep;
 	ssize_t bytes;
 	uint64_t tmp;
 	int iter, lsb_set_pos = 0, pos;
 
-	comp_data = (struct util_coll_join_comp_data *)comp->data;
+	comp_data = (struct util_coll_comp_data *)comp->data;
```
Comment (shefty) [inline prov/util/src/util_coll.c]: This is changing existing code, but if comp->data is a pointer to a struct, can we just define it as that, rather than a void *?

File: prov/util/src/util_coll.c [hunk 30]
```diff
@@ -856,28 +851,28 @@ int ofi_av_set(struct fid_av *av, struct fi_av_set_attr *attr,
...
 	struct util_coll_comp_item *comp_item;
 	int ret;
 
+	coll_mc = (struct util_coll_mc*) ((uintptr_t) coll_addr);
 	comp_item = calloc(1, sizeof(*comp_item));
 	if (!comp_item)
 		return -FI_ENOMEM;
 
-	comp_item->data = (void *) calloc(1, sizeof(uint32_t));
+	comp_item->data = calloc(1, sizeof(uint32_t));
```
Comment (shefty) [inline prov/util/src/util_coll.c]: Maybe I should have kept reading before posting that last comment.  Maybe data can be a union?  This is a weird calloc, because it's not needed (data can store a u32) and doesn't make it clear why we want a u32 value...

Maybe we do a wholesale replacement of the collective code?

**Comment by ooststep** [comment]
going to close this pull request and re-open with a more robust rework of the initial collective implementation.
